### PR TITLE
[chore] use kv.KVStore also for S3 storage

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
This should solve #1105

It replaces our current S3 storage driver with `kv.KVStore` using `storage.S3` under the hood.

In time we can move to remove the now fairly-redundant `storage.Driver` interface.

Shoutout to @theSuess who was also working on this at the same time :grinning:

### CODE REVIEW

**Comment by NyaaaWhatsUpDoc** [comment]
Thank you @theSuess for the `go-store` fix!

File: internal/storage/storage.go [hunk 1]
```diff
@@ -52,20 +52,38 @@ type Driver interface {
...
+		proxy := config.GetStorageS3Proxy()
+
+		s3, err := storage.OpenS3(endpoint, bucket, &storage.S3Config{
+			CoreOpts: minio.Options{
+				Creds:  credentials.NewStaticV4(access, secret, ""),
+				Secure: secure,
+			},
+			GetOpts:      minio.GetObjectOptions{},
+			PutOpts:      minio.PutObjectOptions{},
+			PutChunkSize: 2 * 1024 * 1024, // 2MiB
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/storage/storage.go:69]: Ah interesting, nice catch! I think I'm realising now why my previous testing hadn't managed to catch these issues (they were only hitting the case of ReaderSize, not actually chunking things)

**Comment by theSuess** [comment]
I've rolled this out on my test instance and while it still allocates more memory than the uploaded file, it's less than previously.

image

**Comment by NyaaaWhatsUpDoc** [comment]
> I've rolled this out on my test instance and while it still allocates more memory than the uploaded file, it's less than previously.
> 
> image

You see from the `go-store` code it only allocates that (5MiB as we have it set) write buffer once, and reuses it for each chunk. I wonder if this means minio is still heavy on memory allocation even when chunking :thinking:

**Comment by NyaaaWhatsUpDoc** [comment]
> Good job nerds, will review now. Is this ready to merge if it lgtm?

yep go for it!

**Comment by matthewp** [comment]
Should https://github.com/superseriousbusiness/gotosocial/issues/1105 have been closed because of this?

**Pull request merged**